### ISSUE TITLE
Design: Update Settings Dialog to allow for better scaling

### ISSUE THREAD

**Issue by thenadj** [MozillaReality/FirefoxReality#638]
## Expected Behavior
Settings dialog doesn't scale very well. Right now most features are in Developer Options. 

## Possible Solution
Design a better way to allow better discovery of features and scaling.

**Comment by thenadj** [comment]
@avrignaud Noted. Still adding DE-10 (Settings) for feedback. We're trying to understand how to organize settings and what goes where. Please review and send feedback. 
[DE-10 Settings_002.pdf](https://github.com/MozillaReality/FirefoxReality/files/2485502/DE-10.Settings_002.pdf)

**Comment by keianhzo** [comment]
Which are the options for 1.1?

- Speech Data Collection
- Environment
- Crash Reporting
- Developer Options
- Privacy
- Report An Issue
- Controller Options
- Display

Is that right?

Any particular order?

**Comment by avrignaud** [comment]
I think the Settings options for 1.1 are:
- Language 
- Speech (broken out into permission for collecting voice samples, and perhaps where we allow user to choose preferred voice search language? Latter might be better in Language though?)
- Environment
- Crash Reporting
- Developer Options
- Privacy & Security (think we remove the "Privacy" solo entry?)
- Controller Options
- Report an Issue

I don't think we need Search yet - since we still only have one default search engine, nor Account.

@larsbergstrom @bluemarvin Do we have any opinion on whether we should be asking GV to enable about:preferences support (so design can create a set of pref web pages to use), or whether we should invest time in extending the current settings UI?

**Comment by thenadj** [comment]
Attached is the updated UI Spec for Settings. Ready for review. 
[Uploading UIS-16 Settings_013.pdf…]()

**Comment by thenadj** [comment]
See this issue for additional content to be added to settings: https://github.com/MozillaReality/FirefoxReality/issues/634

**Comment by keianhzo** [comment]
@thenadj I don't seem to find the UIS-16 Settings_013.pdf. The link above redirects to another issue so I'm not sure which one is the final version of the spec.

**Comment by thenadj** [comment]
[UIS-16 Settings_013.pdf](https://github.com/MozillaReality/FirefoxReality/files/2530689/UIS-16.Settings_013.pdf)
Here you go @keianhzo

**Comment by thenadj** [comment]
Please use this spec for updated visual styles and more consistency across all dialogs. 
[UIS-16 Settings_014.pdf](https://github.com/MozillaReality/FirefoxReality/files/2535563/UIS-16.Settings_014.pdf)

**Comment by thenadj** [comment]
Attached is a document with visual bugs for Settings. Looks like some items never got implemented as specd, others were added without design seeing it and having the opportunity to create UI Specs. 

Here is this list of issues. Please see attached document for details. 

- There is no title for the dialog box.
- The RESET ALL line is not in the same container as the other options
- The RESET button is not aligned with the other buttons.
- The scrollbar is the incorrect color.
- There is a title line for languages that should not be there.
- Unnecessary margin for language options.
- Text Field is not editable and has some scrolling effect that cuts the numbers o.
- Toggle ON/OFF text is below the toggle, rather than on the left side of it.

[Bug-17 Settings_002.pdf](https://github.com/MozillaReality/FirefoxReality/files/2590918/Bug-17.Settings_002.pdf)

**Comment by caseyyee** [comment]
Fixed with https://github.com/MozillaReality/FirefoxReality/pull/827

**Comment by cvan** [comment]
The buttons and all the problematic L10n strings were fixed in #827, but let's keep this open for the enhancements for consistent Settings UI panes and the additional Settings (per the proposed Settings IA breakdown). We should file more granular bugs though, but until then let's reopen. The bulk of the work has been done here. Anyone feel free to close and file follow-up issues as needed. Thanks, everyone!

**Comment by philip-lamb** [comment]
Reference: 
[UIS-16_Settings_016.pdf](https://github.com/MozillaReality/FirefoxReality/files/2909833/UIS-16_Settings_016.pdf)
https://trello.com/c/P52DGGmp/88-uis-16-settings

**Comment by MortimerGoro** [comment]
@philip-lamb How is the permission list in settings used? Is it only for web permissions?

AFAIK there is not a programmatically api to revoke Android permissions. Users can only revoke granted permissions from system Settings -> Applications -> your_application -> Permissions.

Some scenarios

- If a user grants microphone access when clicking voice search, what happens if it later disables the permission on settings? The native Android permission is still granted. Does that mean we need to add another custom dialog to prompt that permission if the user clicks voice search again?
- If a permission is disabled and a web requests that permission, should we revoke the permission request automatically or still show the prompt? Web based permission are granted per URL so it wouldn't make any difference enabling or disabling a global permission if we still need to show the prompt in both cases.

For v1.1.3 we could use that UI for web based permissions. If a permission is disabled on the settings list we automatically revoke the permission request and no prompt is shown when a page requests it. If the permission is enabled the prompt is shown as it does in the current release now.

**Comment by philip-lamb** [comment]
Good observations. I don't think the design spec really took into account per-site permissions, but I agree that we might need to provide extra control to the user there.

As regards providing control over permissions that are actually system-permissions; I've seen 3 models for this:

1. If the permission has not been granted, it shows a button with a label, e.g. "Enable location access" which when pressed triggers the system dialog asking for permission. If/when permission has been granted, instead of the button it shows a label with an explanation like e.g. "This app uses location to enable search results related to your surroundings." If the user wants to disable it, they have to do that in system permissions.

2. Shows buttons in-app with the current system permission, but taps on the button cause the system permissions setting app to open on the appropriate page.

3. Show buttons in-app with the current system permission. If it's off and you toggle it to on, it triggers the system permission dialog. If it's on and you toggle it to off, it can't actually revoke the system permission, so this state must be maintained separately (and there's potential for user confusion since now the system permission could be on while the in-app permission could be off). 

My preference is option 1. On entry to the menu, we can query the current state with checkSelfPermission and then enable/disable buttons and labels to show the current state.

**Comment by philip-lamb** [comment]
Depends on #214

**Comment by philip-lamb** [comment]
- [ ] Move Environments selector out of Developer options
- [ ] Move Pointer options out of Developer options
- [ ] Align remaining items with settings hierachy in UIS.

**Comment by MortimerGoro** [comment]
@philip-lamb I implemented similar thing to option 1 for now. For the future we can consider to add per-site permissions
